source/extensions/http/injected_credentials/oauth2/oauth_client.cc

```diff
@@ -0,0 +1,110 @@
+#include "source/extensions/http/injected_credentials/oauth2/oauth_client.h"
+
+#include <chrono>
+
+#include "envoy/http/async_client.h"
+#include "envoy/http/message.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "source/common/common/empty_string.h"
+#include "source/common/common/fmt.h"
+#include "source/common/common/logger.h"
+#include "source/common/http/message_impl.h"
+#include "source/common/http/utility.h"
+#include "source/common/protobuf/message_validator_impl.h"
+#include "source/common/protobuf/utility.h"
+#include "source/extensions/http/injected_credentials/oauth2/oauth_response.pb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace InjectedCredentials {
+namespace OAuth2 {
+
+namespace {
+constexpr const char* GetAccessTokenBodyFormatString =
+    "grant_type=client_credentials&client_id={0}&client_secret={1}";
+constexpr const char* GetAccessTokenBodyFormatStringWithScopes =
+    "grant_type=client_credentials&client_id={0}&client_secret={1}&scope={2}";
+
+} // namespace
+
+OAuth2Client::GetTokenResult OAuth2ClientImpl::asyncGetAccessToken(const std::string& client_id,
+                                                                   const std::string& secret,
+                                                                   const std::string& scopes) {
+  const auto encoded_client_id = Envoy::Http::Utility::PercentEncoding::encode(client_id, ":/=&?");
+  const auto encoded_secret = Envoy::Http::Utility::PercentEncoding::encode(secret, ":/=&?");
+
+  Envoy::Http::RequestMessagePtr request = createPostRequest();
+  const std::string body =
+      fmt::format(GetAccessTokenBodyFormatString, encoded_client_id, encoded_secret);
+  if (!(scopes.empty())) {
+    const auto encoded_scopes = Envoy::Http::Utility::PercentEncoding::encode(scopes, ":/=&?");
+    const std::string body = fmt::format(GetAccessTokenBodyFormatStringWithScopes,
+                                         encoded_client_id, encoded_secret, encoded_scopes);
+  }
+  request->body().add(body);
+  request->headers().setContentLength(body.length());
+  return dispatchRequest(std::move(request));
+}
+
+OAuth2Client::GetTokenResult
+OAuth2ClientImpl::dispatchRequest(Envoy::Http::RequestMessagePtr&& msg) {
+  const auto thread_local_cluster = cm_.getThreadLocalCluster(uri_.cluster());
+  if (thread_local_cluster != nullptr) {
+    thread_local_cluster->httpAsyncClient().send(
+        std::move(msg), *this,
+        Envoy::Http::AsyncClient::RequestOptions().setTimeout(
+            std::chrono::milliseconds(PROTOBUF_GET_MS_REQUIRED(uri_, timeout))));
```

Comment:
What will happen if the client is destructed because the xDS updating and then the a response is received from remote server?
I think you may need to store the async request here and use the result to cancel any pending request when the client is destructed.

Reply:
thanks for pointing out. Addressed it.